File: app/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/activities/fragments/investigation/mapsmenu/MapMenuFragment.java
package com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.mapsmenu;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.AssetManager;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.GridView;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.navigation.Navigation;

import com.TritiumGaming.phasmophobiaevidencepicker.R;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.InvestigationFragment;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.mapsmenu.mapdisplay.data.models.MapListModel;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.mapsmenu.mapdisplay.data.models.MapModel;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.mapsmenu.mapdisplay.io.MapFileIO;
import com.TritiumGaming.phasmophobiaevidencepicker.activities.fragments.investigation.mapsmenu.mapdisplay.io.MapFileReader;
import com.TritiumGaming.phasmophobiaevidencepicker.data.utilities.BitmapUtils;

import java.io.IOException;

/**
 * MapMenuFragment class
 *
 * @author TritiumGamingStudios
 */
public class MapMenuFragment extends InvestigationFragment {

    private MapListModel mapListModel;

    /*
    public MapMenuFragment() {
        super(R.layout.fragment_mapmenu);
    }
    */

    @Nullable
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {

        return inflater.inflate(R.layout.fragment_mapmenu, container, false);
    }

    @SuppressLint({"UseCompatLoadingForDrawables", "ResourceType"})
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {

        super.onViewCreated(view, savedInstanceState);

        Log.d("Maps", "starting");
        try {
            MapFileIO mapFileIO = new MapFileIO();
            MapFileReader reader = mapFileIO.reader;
            if (getActivity() != null) {
                AssetManager assets = getActivity().getAssets();
                mapFileIO.readFile(assets.open("maps.json"), reader);
                mapListModel = new MapListModel(reader.mapsWrapper);
                mapListModel.orderRooms();
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        // INITIALIZE VIEWS
        AppCompatImageView backgroundImage = view.findViewById(R.id.imageView);

        // BACKGROUND IMAGE
        BitmapUtils bitmapUtils = new BitmapUtils();
        bitmapUtils.setResource(R.drawable.icon_map_sm);
        backgroundImage.setImageBitmap(bitmapUtils.compileBitmaps(getContext()));

        GridView gridView = view.findViewById(R.id.grid_maps);
        CustomAdapter customAdapter = new CustomAdapter(
                mapMenuViewModel.getMapNames(),
                mapMenuViewModel.getMapThumbnails());
        gridView.setAdapter(customAdapter);
        gridView.setOnItemClickListener((parent, itemView, position, id) -> {
            System.gc();
            if (mapMenuViewModel != null) {
                mapMenuViewModel.setCurrentMapData(position);
            }

            MapModel mapModel = mapListModel.getMapById(position);
            if(mapModel != null) {
                mapMenuViewModel.setCurrentMapModel(mapModel);
            }

            navigateToBasicMapView(itemView);
        });

    }

    private void navigateToBasicMapView(View view) {
        Navigation.findNavController(view).navigate(R.id.action_mapMenuFragment_to_mapViewerFragment);
    }

    public void initNavListeners(View lstnr_navLeft,
                                  View lstnr_navMedLeft,
                                  View lstnr_navCenter,
                                  View lstnr_navMedRight,
                                  View lstnr_navRight,
                                  AppCompatImageView icon_navLeft,
                                  AppCompatImageView icon_navMedLeft,
                                  AppCompatImageView icon_navCenter,
                                  AppCompatImageView icon_navMedRight,
                                  AppCompatImageView icon_navRight) {

        if(lstnr_navLeft != null) {
            ((View)lstnr_navLeft.getParent()).setVisibility(View.VISIBLE);
            icon_navLeft.setImageResource(R.drawable.icon_evidence);
            lstnr_navLeft.setOnClickListener(v -> Navigation.findNavController(v)
                    .popBackStack()
            );
        }

        if(lstnr_navMedLeft != null) { }

        if(lstnr_navCenter != null) { }

        if(lstnr_navMedRight != null) { }

        if(lstnr_navRight != null) { }

    }

    @Override
    public void softReset() {

    }

    public class CustomAdapter extends BaseAdapter {

        private String[] mapNames;
        private @DrawableRes int[] images;
        private LayoutInflater layoutInflater;

        public CustomAdapter(String[] mapNames, @DrawableRes int[] images) {
            if(getContext() == null) {
                return;
            }
            this.mapNames = mapNames;
            this.images = images;

            layoutInflater =
                    (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        }

        @Override
        public int getCount() {
            return images.length;
        }

        @Override
        public Object getItem(int position) {
            return null;
        }

        @Override
        public long getItemId(int position) {
            return 0;
        }

        @Override
        public View getView(int i, View newView, ViewGroup parent) {

            if(newView == null) {
                newView = layoutInflater.inflate(R.layout.item_mapmenu_map, parent, false);
            }

            AppCompatTextView textView = newView.findViewById(R.id.label_mapName);
            AppCompatImageView imageView = newView.findViewById(R.id.image_map);

            textView.setText(mapNames[i].split(" ")[0]);
            imageView.setImageResource(images[i]);

            return newView;
        }
    }

    /**
     * Enforces garbage collection on low memory
     */
    @Override
    public void onLowMemory() {
        System.gc();

        super.onLowMemory();
    }
}

File: app-legacy/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/presentation/ui/activities/investigation/mapsmenu/MapMenuFragment.kt
package com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation.mapsmenu

import android.annotation.SuppressLint
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.BaseAdapter
import android.widget.GridView
import android.widget.Toast
import androidx.annotation.DrawableRes
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.AppCompatTextView
import androidx.navigation.Navigation.findNavController
import com.google.common.primitives.Ints
import com.tritiumgaming.phasmophobiaevidencepicker.R
import com.tritiumgaming.phasmophobiaevidencepicker.domain.model.maps.io.MapFileIO
import com.tritiumgaming.phasmophobiaevidencepicker.domain.model.maps.map.MapListModel
import com.tritiumgaming.phasmophobiaevidencepicker.presentation.ui.activities.investigation.InvestigationFragment

/**
 * MapMenuFragment class
 *
 * @author TritiumGamingStudios
 */
class MapMenuFragment : InvestigationFragment() {

    private var mapListModel: MapListModel? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_mapmenu, container, false)
    }

    @SuppressLint("UseCompatLoadingForDrawables", "ResourceType")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        try { mapListModel = readMapsDataFromFile() }
        catch (e: Exception) { e.printStackTrace() }

        val gridView = view.findViewById<GridView>(R.id.grid_maps)

        mapListModel?.let { mapListModel ->
            mapMenuViewModel?.let { mapMenuViewModel ->
                val gridViewAdapter = GridViewAdapter(
                    mapListModel.shortenedMapNames.toTypedArray<String?>(),
                    Ints.toArray(mapMenuViewModel.mapThumbnails)
                )
                gridView.adapter = gridViewAdapter
                gridView.onItemClickListener =
                    AdapterView.OnItemClickListener {
                        _: AdapterView<*>?, itemView: View, position: Int, _: Long ->
                        mapListModel.let { mapListModel ->
                            System.gc()

                            mapMenuViewModel.currentMapIndex = position
                            mapMenuViewModel.currentMapModel = mapListModel.getMapById(position)

                            navigateToMapView(itemView)
                        }
                    }
            } ?:  Toast.makeText(requireContext(),
                getString(R.string.alert_error_generic), Toast.LENGTH_LONG).show()
        } ?: Toast.makeText(requireContext(),
            getString(R.string.alert_error_generic), Toast.LENGTH_LONG).show()
    }

    @Throws(Exception::class)
    private fun readMapsDataFromFile(): MapListModel? {
        val mapFileIO = MapFileIO()
        val reader = mapFileIO.mapFileReader

        mapFileIO.readFile(requireActivity().assets, getString(R.string.mapsJson), reader)
        mapListModel = MapListModel(reader.worldMapDeserializer)
        mapListModel?.orderRooms()

        return mapListModel
    }

    private fun navigateToMapView(view: View) {
        //findNavController(view).navigate(R.id.mapViewerFragment)
        try {
            findNavController(view).navigate(R.id.action_mapMenuFragment_to_mapViewerFragment)
        } catch (e: IllegalArgumentException) {
            e.printStackTrace()
        }
    }

    inner class GridViewAdapter(
        mapNames: Array<String?>, @DrawableRes images: IntArray
    ) : BaseAdapter() {

        private var mapNames = arrayOfNulls<String>(0)
        @DrawableRes private var images = IntArray(0)

        init {
            if (mapNames.size == images.size) {
                this.mapNames = mapNames
                this.images = images
            } else {
                Log.d("MapMenu", "Name count: ${mapNames.size} Images count: ${images.size}")
            }
        }

        override fun getCount(): Int { return images.size }

        override fun getItem(position: Int): Any? { return null }

        override fun getItemId(position: Int): Long { return 0 }

        override fun getView(i: Int, passedNewView: View?, parent: ViewGroup): View {
            val newView: View = passedNewView?.let { passedNewView }
                ?: layoutInflater.inflate(R.layout.item_mapmenu_map, parent, false)

            val textView = newView.findViewById<AppCompatTextView>(R.id.label_mapName)
            val imageView = newView.findViewById<AppCompatImageView>(R.id.image_map)

            textView?.text = mapNames[i]
            imageView?.setImageResource(images[i])

            return newView
        }
    }

    /**
     * Enforces garbage collection on low memory
     */
    override fun onLowMemory() {
        System.gc()

        super.onLowMemory()
    }

    override fun reset() {
    }

    override fun saveStates() {
    }
}